LXStudio-IDE/src/main/java/titanicsend/output/TESacnOutput.java

```diff
@@ -0,0 +1,106 @@
+import java.net.UnknownHostException;
+import java.util.*;
+
+public class TESacnOutput {
+  private static class EdgeEntry {
+    TEEdgeModel edge;
+    int universeNum;
+    int strandOffset;
+    public EdgeEntry(TEEdgeModel edge, int universeNum, int strandOffset) {
+      this.edge = edge;
+      this.universeNum = universeNum;
+      this.strandOffset = strandOffset;
+    }
+  }
+  String ipAddress;
+  static Map<String, TESacnOutput> ipMap = new HashMap<>();
+  private final List<EdgeEntry> edgeEntries;
+  private boolean activated;
+  private HashMap<Integer,Integer> deviceLengths;
+  private StreamingACNDatagram outputDevice;
+
+  private TESacnOutput(String ipAddress) {
+    this.ipAddress = ipAddress;
+    this.edgeEntries = new ArrayList<>();
+    this.activated = false;
+    this.outputDevice = null;
+  }
+
+  public static TESacnOutput getOrMake(String ipAddress) {
+    if (!ipMap.containsKey(ipAddress)) {
+      ipMap.put(ipAddress, new TESacnOutput(ipAddress));
+    }
+    return ipMap.get(ipAddress);
+  }
+
+  public static void register(TEEdgeModel edge, String ipAddress, int deviceNum, int strandOffset) {
+    assert deviceNum >= 1;
+    assert deviceNum <= 4;
+    assert strandOffset >= 0;
+    TESacnOutput output = getOrMake(ipAddress);
+    assert !output.activated;
+    output.edgeEntries.add(new EdgeEntry(edge, deviceNum, strandOffset));
+  }
+
+  // Sort by device number, then by strand offset
+  private static class SortEdgeEntries implements Comparator<EdgeEntry> {
+    public int compare(EdgeEntry a, EdgeEntry b) {
+      if (a.universeNum != b.universeNum) {
+        return a.universeNum - b.universeNum;
+      } else {
+        return a.strandOffset - b.strandOffset;
+      }
+    }
+  }
+
+  private void activate(LX lx) {
+    assert !this.activated;
+    this.deviceLengths = new HashMap<>();
+    this.edgeEntries.sort(new SortEdgeEntries());
+    int currentUniverseNum = 0;
+    int currentStrandOffset = -1;
+
+    StringBuilder logString = new StringBuilder("sACN " + this.ipAddress + ": ");
+    for (EdgeEntry edgeEntry : this.edgeEntries) {
+      int edgeLength = edgeEntry.edge.points.length;
+      if (edgeEntry.universeNum > currentUniverseNum) {
+        currentUniverseNum = edgeEntry.universeNum;
+        currentStrandOffset = 0;
+        String deviceSummary = "#" + currentUniverseNum + " ";
+        logString.append(deviceSummary);
+      }
+      assert edgeEntry.universeNum == currentUniverseNum;
+      assert edgeEntry.strandOffset == currentStrandOffset;
+      String edgeSummary = "[" + currentStrandOffset + ":Edge_" + edgeEntry.edge.id() + "=" + edgeLength + "] ";
+      logString.append(edgeSummary);
+      currentStrandOffset += edgeLength;
+      this.deviceLengths.put(currentUniverseNum, currentStrandOffset);
+
+      // FIXME: We're not doing anything with the strand offset.
+      this.outputDevice = new StreamingACNDatagram(lx, edgeEntry.edge, currentUniverseNum);
```

Comment:
I believe you'll want to use the `StreamingACNDatagram(LX lx, int[] indexBuffer, int universeNumber)` constructor instead of the one that passes in the whole model. `indexBuffer` here is an array of the `index` of each `LXPoint` that should be controlled by the output. The constructor you have here actually uses [`LXModel.toIndexBuffer`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/model/LXModel.java#L703) under the hood.

Reply:
That way, you can combine the indices of each edge covered by this output into a single `indexBuffer`.

Reply:
That all sounds right, but here's what's weird: the code above is only called once, at initialization time, and then somehow LX knows to keep pulling the model data out and sending datagrams with it, even though I'm not touching any of these data structures in any of TE's own loop code. I'm guessing there's some indirection going on, but I haven't had time to trace it all out with the debugger.

To put it another way, after we build our own concatenated indexBuffer, what do we do with it next? And is this done at initialization time or on every loop?

Reply:
Here's how I understand it. For simplicity, imagine we have a strip spanning 2 edges with a total of 4 pixels with the indexes `{0, 1, 4, 5}` (where by index, I mean `LXPoint.index` - an integer unique to each pixel in the top-level model that points into a global `colors` buffer). Now we create an sACN output like this: `StreamingACNDatagram(lx, new int[]{0, 1, 4, 5})`. The LX engine calls [`LXDatagram.onSend`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/output/LXDatagram.java#L198) on every loop, which in turn calls [`LXBufferOutput.updateDataBuffer`](https://github.com/heronarts/LX/blob/master/src/main/java/heronarts/lx/output/LXBufferOutput.java#L110) to insert the current colors for the pixels controlled by our output into the buffer that stores the data we send over the wire. The fact that this code is only called once is fine as long as our LXPoint indices don't change later.

To maybe clear up some confusion that I first had when reading the code, `StreamingACNDatagram` is not just a single static datagram - it's actually an output that continuously transmits data in the form of sACN datagrams.